src/backend/distributed/commands/utility_hook.c

```diff
@@ -1694,7 +1703,29 @@ IsStatementSupportedFromNonMainDb(Node *parsetree)
 	{
 		if (type == NonMainDbSupportedStatements[i].statementType)
 		{
-			return true;
+			if (NonMainDbSupportedStatements[i].supportedObjectTypes == NULL)
+			{
+				return true;
+			}
+			else
+			{
+				if (type == T_GrantStmt)
+				{
+					GrantStmt *stmt = castNode(GrantStmt, parsetree);
+
+					/* check if stmt->objtype is in supportedObjectTypes */
+					for (int j = 0; j <
+						 NonMainDbSupportedStatements[i].supportedObjectTypesSize; j++)
+					{
+						if (stmt->objtype ==
+							NonMainDbSupportedStatements[i].supportedObjectTypes[j])
+						{
+							return true;
+						}
+					}
+					return false;
+				}
```

Comment:
To me, this block feels a bit hard to maintain. Would you like to cherry-pick the changes from https://github.com/citusdata/citus/commit/6d73c8381cf44d248229a5d5f9e1ea1632f82058 instead?

This one makes use of some function pointers to avoid clutter in IsStatementSupportedFromNonMainDb.

Reply:
Apart from that, those non-main db related code definitely need to be moved into a new file soon but I'm refraining from doing this now to avoid further merge conflicts.

However, we should do that maybe after merging currently active PRs touching this code-path.

Reply:
Totally agree with you, after maturization of this structure, we can easily move it to another file/files

Reply:
I like the function pointers solution. I'll cherry-pick and make required changes if it is required
Thanks

Reply:
Cherry-picked. Thanks for the contribution